### ISSUE TITLE
Open in new tab

### ISSUE THREAD

**Issue by Luticus** [nextcloud/external#79]
It's really great that external sites has the ability to handle sites that can't be embedded, however, it would be really great if there were an option to launch them in a new tab instead of closing/overwriting my Nextcloud tab.

**Comment by nickvergessen** [comment]
This is the current decision.
Use CTRL+click or right-click to open the link in a new tab.

**Comment by Luticus** [comment]
Yep, that's what I've been doing up till now.  Was hoping for a better more integrated and automatic method but I suppose you can't win them all.  Thanks for checking into it.

**Comment by Luticus** [comment]
App is great, btw, loving the per-site access groups/users thing.  Made my life a lot better with that alone so I'll take it.  :)

**Comment by vinzv** [comment]
Would it be possible to get some checkbox into admin's settings to enable e.g. target="_blank" on a per-link basis?

**Comment by nickvergessen** [comment]
The problem is, that the API of the server does not have this option.

**Comment by vinzv** [comment]
I see. We can live with the current behaviour but I'd like to add my thoughts on that so you maybe can consider it for the future:

Default behaviour is embedding sites. If that's not possible, the target page is loaded solely but "overwrites" the nextcloud page. I think it would be more intuitive to have it loaded in another tab/new window instead of killing the user's current view.

**Comment by zippytiff** [comment]
Hi,

Any eta on the enhancement ? keen to get this out to users, but they wont cope with any work arounds

ZT

**Comment by jancborchardt** [comment]
Not sure how this decision was taken, but I agree with what people mentioned here. If a site can’t be embedded, then clicking it should _not_ overwrite the Nextcloud tab but instead open the site in a new tab as it’s external.

cc @nickvergessen @juliushaertl

**Comment by KnownNexus** [comment]
Hi,
I'd just like to clarify the position on this,
i've been developing a change to make it so:
left clicking, ctrl-clicking and right-click: open new tab - all open the file in a new tab

ctrl-click was previously bound to "single select"
and left click resulted in opening the file in the current window (Not ideal for my use case)


Is this something you guys would be interested in?
and if you have any issues, how would you like it changed to make it acceptable?

**Comment by juliushaertl** [comment]
> Not sure how this decision was taken, but I agree with what people mentioned here. If a site can’t be embedded, then clicking it should not overwrite the Nextcloud tab but instead open the site in a new tab as it’s external.

Sounds good to me, but only for non-embedded sites

> left clicking, ctrl-clicking and right-click: open new tab - all open the file in a new tab

Right click should still just open the browser menu, no need to overwrite the default behavior here.

**Comment by KnownNexus** [comment]
> Right click should still just open the browser menu, no need to overwrite the default behavior here.
Sorry just to clarify, right click opens the context menu, i haven't modified any functionality in that case, i was just confirming that it did still open a new tab

**Comment by jmacarthur** [comment]
The current behaviour seems to be a mix of two common behaviours:

For a multiple-selection list box, single click selects one item and deselects others, shift-click selects a range, and ctrl-click toggles one item on or off.
For a HTML link, single click opens in the same window, shift-click opens in new window, and ctrl-click opens in a new tab.

These are just what I consider to be normal behaviour, your browser and experience with UI toolkits may be different of course. But if you agree with the above models, shouldn't we pick one rather than mixing them?

**Comment by Luticus** [comment]
When I originally suggested this, what I thought of was a check box in settings where you declare the external site that says something like "open this site in new tab" and then when you check that box and a user clicks the site icon rather than opening in the same tab and clobbering the users Nextcloud session it opens in a new tab per the browsers settings.  I'm not particularly interested in the crtl + click behavior as I'm fairly certain like 99% of users probably don't realize that's a thing.  Right click menu options could be interesting.  Right click on the icon and you get a menu that says open in new tab or something like that would be interesting, so that other apps could be opened easily in new tabs, like talk, for instance.  I'm most interested in external sites though.  I embed emby and others in nextcloud and they don't behave right when they are loaded as an app.  Full screening a movie doesn't work in emby for instance.  So I set them up to clobber nextcloud.  Would be really nice if instead it just opened in a new tab.

**Comment by jancborchardt** [comment]
@KnownNexus could you open a pull request with just 1 change as described above? That would likely fix the issue, without any need for modification or learning patterns on people’s part:

> If a site can’t be embedded, then clicking its icon should not overwrite the Nextcloud tab but instead open the site in a new tab.

Does that work? Thanks a lot! :)

-------------------



> When I originally suggested this, what I thought of was a check box in settings where you declare the external site that says something like "open this site in new tab" and then when you check that box and a user clicks the site icon rather than opening in the same tab and clobbering the users Nextcloud session it opens in a new tab per the browsers settings. I'm not particularly interested in the crtl + click behavior as I'm fairly certain like 99% of users probably don't realize that's a thing.

@Luticus The issue is that the majority of people would not realize that a setting for this exists and then be none the wiser either. :) This needs to simply happen automatically, as said above.

**Comment by KnownNexus** [comment]
> @KnownNexus could you open a pull request with just 1 change as described above? That would likely fix the issue, without any need for modification or learning patterns on people’s part:
> 
> > If a site can’t be embedded, then clicking its icon should not overwrite the Nextcloud tab but instead open the site in a new tab.

What about things like libreoffice? I think it would be good to have that kind of thing open in a new tab, rather than embed in the current one, as that means you aren't losing your place in the file system if you close it.
(Many peoples reaction is to close tabs rather than clicking to return to the previous page)
Something similar to the way google drive opens its docs

**Comment by nickvergessen** [comment]
> Not sure how this decision was taken…

By technology. The app can not influence the core's navigation handling to tell it to do this.
Either this needs to be extended on the server side. Or we even hard code it there and when the target is not starting with the nextcloud url, we always open in a new tab.

**Comment by jancborchardt** [comment]
> What about things like libreoffice? I think it would be good to have that kind of thing open in a new tab, rather than embed in the current one, as that means you aren't losing your place in the file system if you close it.

This issue here is about the "External" app, for embedding sites. It’s not about Collabora Online. :) (And Collabora Online doesn’t make you lose your place in the file system, as when you close the Collabora view you are back in Files where you were before.)

**Comment by jmacarthur** [comment]
When editing two or more documents, say to copy text from one into another, it's useful to have the file list remain. Otherwise, you would have to open two instances of the same file window before clicking different files in each. If we have to do that via right-click, that will be fine. Ctrl-click to open in a new tab would be nice, but isn't essential.

**Comment by Luticus** [comment]
@jancborchardt 

> The issue is that the majority of people would not realize that a setting for this exists and then be none the wiser either. :) This needs to simply happen automatically, as said above.

Not sure I follow you here.  Maybe we are talking about 2 different things.  I'm talking about taking the "default" behavior out of the hands of the users entirely.  So as an admin when you are declaring a url as an external site you either add a check box that says "open this site in new tab", or even just modify the behavior of the "this site does not allow embedding" checkbox, which honestly works fine for me.  As an admin it's easier to just set the defaults that make the most sense for users rather than explaining, "no on this icon you need to right click and and this one you need to ctrl+click".  Just, click the icon and if it works better in a new tab, it opens in a new tab.  

![ncexternalsites_settings](https://user-images.githubusercontent.com/7807109/59924098-f1422480-9402-11e9-9b04-cc4cf0042d8f.png)

Maybe even display a popup message, similar to when you send a message over 1000 characters in the talk app, for the user that says that the item opened in a new tab.  This way if their browser is configured to not shift immediately to a new tab the user can tell that the icon did something.  Might not seem important but as a user who keeps lots of tabs open and doesn't let the browser transition to them immediately on open, I can tell you that it's sometimes easy to get lost and not realize a link opened in a new tab.

**Comment by TP75** [comment]
> > Not sure how this decision was taken, but I agree with what people mentioned here. If a site can’t be embedded, then clicking it should not overwrite the Nextcloud tab but instead open the site in a new tab as it’s external.
> 
> Sounds good to me, but only for non-embedded sites

@nickvergessen Please accept my apologies for being insistent or if I may misunderstand. IMHO this delay and long discussion is confusing. What is the status?

To my current knowledge an external site does not get a redirect as expected and as is documented.

Please make true redicts possible in due time. Redirects for entries marked as 'redirect - not allow embedding' and clicking the resp icon should not overwrite the Nextcloud tab but instead open the site in a new tab as it’s external.

**Comment by jancborchardt** [comment]
Yes, long discussion is confusing and adding more replies won’t make it better. If there is any progress, this issue will be updated. Contributions are very welcome and what was said above still stands:

https://github.com/nextcloud/external/issues/79#issuecomment-503501946

**Comment by nooblag** [comment]
![Capture](https://user-images.githubusercontent.com/1122344/76606134-dc478b80-6565-11ea-927d-f978dd2b49cf.PNG)

**Comment by nickvergessen** [comment]
No we will not add a user specified target option.

**Comment by poltrian** [comment]
A target will be helpful. You cannot explain the right click to every user.
Especially since you do not know beforehand whether the page opens in Nexcloud or in the same tab

**Comment by nickvergessen** [comment]
So you can not explain people aright click, but want to explain them to fill this field with `_blank` etc?

Anyway back to the topic. opening in a new tab based on the redirect should happen and is still on the list, but nothing more.

**Comment by nooblag** [comment]
> So you can not explain people aright click, but want to explain them to fill this field with `_blank` etc?

You're so grumpy! No. Admins would set this in the External Links settings for each link depending on it's other options such as 'position' and 'icon,' etc.

> Anyway back to the topic. opening in a new tab based on the redirect should happen and is still on the list, but nothing more.
That's good news at least. ;)

**Comment by acs-ferreira** [comment]
The problem here is that the link is parsed with JS. So, even a link like

`<a href="https://example.com/" target="_blank"></a>`

or

`<a href="https://example.com/" onclick="window.open(this.href)"></a>`

will not open in a new tab.

**Comment by nooblag** [comment]
Thanks for a better explanation.

Is there an ETA for opening in a new tab based on the redirect setting?

**Comment by jancborchardt** [comment]
Can only repeat what I said above:

> Yes, long discussion is confusing and adding more replies won’t make it better. If there is any progress, this issue will be updated. Contributions are very welcome and what was said above still stands:
> https://github.com/nextcloud/external/issues/79#issuecomment-503501946

And will hereby lock the issue. Further discussion can happen in a pull request which is opened by someone who would like to implement this as discussed. :)
